oci/spec_opts_linux.go

```diff
@@ -26,22 +26,63 @@ import (
 
 	"github.com/containerd/containerd/containers"
 	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
 	"golang.org/x/sys/unix"
 )
 
 // WithHostDevices adds all the hosts device nodes to the container's spec
 func WithHostDevices(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
 	setLinux(s)
 
-	devs, err := getDevices("/dev")
+	devs, err := getDevices("/dev", "")
 	if err != nil {
 		return err
 	}
 	s.Linux.Devices = append(s.Linux.Devices, devs...)
 	return nil
 }
 
-func getDevices(path string) ([]specs.LinuxDevice, error) {
+// WithDevices recursively adds devices from the passed in path and associated cgroup rules for that device.
+// If devicePath is a dir it traverses the dir to add all devices in that dir.
+// If devicePath is not a dir, it attempts to add the single device.
+// If containerPath is not set then the device path is used for the container path.
+func WithDevices(devicePath, containerPath, permissions string) SpecOpts {
+	return func(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
+		devs, err := getDevices(devicePath, containerPath)
+		if err != nil {
+			return err
+		}
+		for _, dev := range devs {
+			s.Linux.Devices = append(s.Linux.Devices, dev)
+			s.Linux.Resources.Devices = append(s.Linux.Resources.Devices, specs.LinuxDeviceCgroup{
+				Allow:  true,
+				Type:   dev.Type,
+				Major:  &dev.Major,
+				Minor:  &dev.Minor,
+				Access: permissions,
+			})
+		}
+		return nil
+	}
+}
+
+func getDevices(path, containerPath string) ([]specs.LinuxDevice, error) {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return nil, errors.Wrap(err, "error stating device path")
+	}
+
+	if !stat.IsDir() {
+		dev, err := deviceFromPath(path)
+		if err != nil {
+			return nil, err
```

Comment:
Do we not need the same logic on L117?

if err == ErrNotADevice {
				continue					continue
			}				}
			if os.IsNotExist(err) {				if os.IsNotExist(err) {
				continue					continue
			}				}

Reply:
maybe we can extract logic from line 76 to line 82 into a method and we can reuse it for logic from line 115 to line 127?

Reply:
This is different here.
We aren't iterating a directory tree, someone specified a device path but the device is not a device.

Reply:
True, the only thing is this seems to lead to an inconsistent behavior of this method - when the caller specifies a single invalid device path, we return error, but when the caller specifies a folder with none but invalid devices inside it, we return empty list of device info.  I wonder if we always return empty list of device info would make it a bit easier for the caller (but I realize this is a private method so I'm totally fine with the current way if you feel my comment doesn't make sense).

Reply:
I see what you mean. I'm not sure what the right thing to do is here.
So we are erroring in all cases where the path is not a device or a directory (or a symlink to a directory, I suppose).

I can see where someone might want a universal function to handle all the things here, but I'm not sure what's best at this layer.
Really anyone can write a function to modify the spec to add devices in whatever way they want. This is really here for cri, and it happens to be more generally useful.
Callers can check the error, too.

Reply:
got you. thanks for explaining.